### PR TITLE
Anonymous helper functions AKA Inline JavaScript expressions

### PR DESCRIPTION
**Pull request by xiphias**
I spent some time having fun implementing this feature, which creates anonymous helper functions.
It works well enough, although better and more unit tests are needed that can test the created functions.
It works with templates and files as well, and works with #{} and !{} syntax and if markups.
The next logical step would be anonymous events (and inlne coffeescript support), but that doesn't belong to this CL.

### CODE REVIEW

**Comment by xiphias** [comment]
This is what simple-todos looks like for me while still working:

simple-todos.jade:

head
  title Todo List

body
  .container
    header
      h1 Todo List (#{Tasks.find({checked: {$ne: true}}).count()})
      label.hide-completed
        input(type="checkbox" checked="#{Session.get('hideCompleted')}")
        | Hide Completed Tasks

      +loginButtons
      if currentUser
        form.new-task
          input(type="text" name="text" placeholder="Type to add new tasks!")
    ul
      each tasks
        +task

task.tpl.jade:

li(class="{{#if checked}}checked{{/if}} {{#if private}}private{{/if}}")
    button.delete &times;
    input.toggle-checked(type="checkbox" checked=checked)
    if this.owner==Meteor.userId()
      button.toggle-private
        if private
          | Private
        else
          | Public
    span.text <strong>#{username}</strong> - #{text}

**Comment by mquandalle** [comment]
Hey @xiphias, thank you for this PR, we finally get a chance to fix issue n° #1 :-) I will have to do a careful review of the code before merging. Could you please add a few comments in the code before I dig in?

Also as you noticed, this feature definitely needs more tests, but we can do that in parallel.

**Comment by xiphias** [comment]
Sure, I'll write some comments.

File: packages/jade-compiler/tests/tests.js [hunk 1]
```diff
@@ -30,3 +30,49 @@ Tinytest.add("JadeCompiler - compile templates", function(test) {
   test.equal(JadeCompiler.compile(template),
   "(function() {\n  return HTML.P(\"hello world\");\n})");
 });
+
+var template3 = wrapInTemplate("hello",
+  ["if helper arg1 arg2",
```
Comment (mquandalle) [inline packages/jade-compiler/tests/tests.js]: I think we can use ES6 template string here, just add the meteor `ecmascript` package and replace `"` by `\``

**Comment by xiphias** [comment]
I added the comments, switched to ES6 and fixed Travis build.

**Comment by xiphias** [comment]
A limitation that I found is that the mixins can't have too complex JavaScript..I guess Jade Lexer needs to be modified for that.

**Comment by darrynten** [comment]
This is excellent! +1

**Comment by xiphias** [comment]
Hey @mquandalle, I added some bug fixes and better tests, could you please take another look at the pull request?

**Comment by mquandalle** [comment]
Sure! Also @dalgard, I haven’t reviewed your fork yet but did you implement something similar? Do you have an opinion on this?

**Comment by dalgard** [comment]
Oh, hi @mquandalle – welcome back! :)

I haven't implemented inline JavaScript expressions – I think it's a really bad idea. As I've written elsewhere, the templating paradigm that Blaze builds upon is a logicless one – it is *not* in my opinion desirable to embed JavaScript in Blaze templates.

**Comment by xiphias** [comment]
I have no problem with having a separate package if you don't like the direction. I don't really like the MVC separation that most systems use, and prefer to write logic and UI at the same place, but I also acknowledge that it's not the mainstream software design philosophy. Reactivity in Meteor partly makes the separation of controller and view unnecessary, and in my view having view and event logic in the jade templates make the code easier to maintain, for a non-designer programmer. It also makes it harder for non-programmer designers, but I don't generally work with designers who modify my templates. So it's up to you to make the decision.

**Comment by dalgard** [comment]
@xiphias: I totally agree with you in the sense that React is a nice addition to Meteor, and that I hope that Blaze 2 will embrace some kind of JSX style. But I don't think these concepts should be brought into Jade/Spacebars.

**Comment by xiphias** [comment]
I'm not really talking about React/JSX, which has it's own disadvantages. Anyways, if the decision by @mquandalle is not having this patch either and other people are not interested, I'll probably just create some new features for myself without going through commenting & preparing for submission again.

**Comment by xiphias** [comment]
@mquandalle I added anonymous events as well, please check it if you like it. I think it simplifies Meteor development a lot. You can check the simple-todos implementation with the new features here in the README:
https://github.com/xiphias/meteor-jade

**Comment by mquandalle** [comment]
I for one appreciate the possibility to use inline JavaScript expressions in my templates. It turns out in a large application there are quite a few places where writing `if x < 10` is more convenient and expressive than creating a dump `isLowerThan10` template helper. I’ve also talked about that with David Greenspan (the author of Spacebars and Blaze) and at least at the time he agreed that inline expressions made sense.

In https://github.com/mquandalle/meteor-jade/issues/1#issuecomment-134219158 I have also expressed the serious difficulties that I encounter while trying to implement inline expressions. So I need to be sure that whatever you’re doing @xiphias this doesn’t introduce too many bugs and side effects and that it is consistent enough from a user point of view. I still have the feeling that some of the JavaScript expression handling implemented here overlays with @dalgard fork (for instance ` element(attribute=helper(posArg keyword='cooler'))`) so I need to figure that out.

File: packages/jade/plugin/handler.js [hunk 2]
```diff
@@ -47,6 +82,20 @@ var getCompilerResult = function (compileStep, fileMode) {
   }
 };
 
+// Generate helper functions using function name - function body map
+//   in the input.
+var templateHelperGen = function (helpers, tplName) {
```
Comment (mquandalle) [inline packages/jade/plugin/handler.js]: `templateHelperGen` and `templateEventsGen` are very similar, wouldn’t be better to use the same underlining function?

File: packages/jade/tests/match.html [hunk 3]
```diff
@@ -118,4 +118,9 @@
 <template name="match-html-specialMarkdownComponent">{{#markdown}}This is my email: <address@example.com>
 {{/markdown}}</template>
 
-<template name="match-html-specialMarkdownComponentWithHelper">{{#markdown}}This is my email: {{email}}{{/markdown}}</template>
+<template name="match-html-specialMarkdownComponentWithHelper">{{#markdown}}This is my email: {{email}}
+{{/markdown}}</template>
+
+<template name="match-html-anonymousIfHelper">{{#if _jade_line257}}hello world{{/if}}</template>
```
Comment (mquandalle) [inline packages/jade/tests/match.html]: Relying on the generated helper name (here `_jade_line257`) is kind of fragile because (the line number might change for example) and I think we shouldn’t rely on it. (How do you handle 2 anonymous helpers on the same line BTW?)

I we should create a second tester function in `match.js` that would abstract the anonymous helpers and consider the test successful for whatever helper name we have as long as the helper itself and its surrounding template is correct.

File: packages/jade-compiler/lib/transpilers.js [hunk 4]
```diff
@@ -42,6 +42,69 @@ var throwError = function (message, node) {
   throw new Error(message);
 };
 
+// Recursively search for anonymous helper functions and
+// record them
+// Arguments:
+// value: the tree created by the transpiler.
+// found: a map from helper function names to
+//    their function body.
+function searchNewArgs(value, found, foundEvents) {
```
Comment (mquandalle) [inline packages/jade-compiler/lib/transpilers.js]: Did you consider using [HTMLTools.Scanner](https://github.com/meteor/meteor/blob/devel/packages/html-tools/scanner.js)?

File: packages/jade-compiler/lib/transpilers.js [hunk 5]
```diff
@@ -58,12 +121,40 @@ _.extend(FileCompiler.prototype, {
...
+      r.bodyHelpers=self.bodyHelpers;
+    if(!_.isEmpty(self.templatesHelpers))
+      r.templatesHelpers=self.templatesHelpers
+    return r;
+  },
+
+  // Generate template helpers for every template.
+  genTemplateHelpers: function(templates) {
+    r={}
+    for (var key in templates) {
```
Comment (mquandalle) [inline packages/jade-compiler/lib/transpilers.js]: Generally speaking `for x in y` should be prohibited because it leaks some keys from the prototypes. I’ve been using underscore helpers in this package (as does most packages in the Meteor ecosystem)

File: packages/jade/plugin/handler.js (hunk 2 above)
Comment (xiphias) [inline packages/jade/plugin/handler.js]: I could, but I would need to pass a generator function to create a higher order function, which would make it harder to read. Still, I did some refactoring to make handler.js simpler.

File: packages/jade-compiler/lib/transpilers.js (hunk 5 above)
Comment (xiphias) [inline packages/jade-compiler/lib/transpilers.js]: I tried not to add extra dependencies, and I didn't know about the underscore dependency when I started extending the code, but I fixed it now.

File: packages/jade/tests/match.html (hunk 3 above)
Comment (xiphias) [inline packages/jade/tests/match.html]: I removed the tests. I believe runtime.js is enough for testing to see if the code works as intended (maybe more test cases could be added there to see if there are important not handled edge cases). What do you think?

File: packages/jade-compiler/lib/transpilers.js (hunk 4 above)
Comment (xiphias) [inline packages/jade-compiler/lib/transpilers.js]: I'm not sure where I could make use of it, as the code I wrote doesn't create extra HTML.

**Comment by xiphias** [comment]
Regarding @dalgard 's fork the main problem with the helpers is that they can't call other helpers. Maybe extending the anonymous helper functions with the helpers would solve this issue, I'll try it.

**Comment by dalgard** [comment]
I have to say that I think it's ***not*** the place for the canonical Jade package to introduce a radical new way of writing Blaze templates. This kind of experimental extension of what people expect from Blaze belongs in its own package.

**Comment by xiphias** [comment]
@dalgard I would be fine with this as well, but in this case I would have spent time learning how to create a package instead of trying to fit in the current package.

@mquandalle Please make a decision on the direction you want your package to go to, as I fixed everything that you asked for, and it's your package. I'm happy to make a fork, but I need you to make a decision.

**Comment by dalgard** [comment]
Just to be clear: I don't suggest you make a fork @xiphias, I suggest you make an extension – a supplementary package to the canonical one.

I probably should have done the same with the extensions from my fork, but, like you, I had made my changes as a pull request and I guess it was easier to simply publish my fork.

**Comment by xiphias** [comment]
@dalgard It would be great to be able to create an extension, but there are no extension hooks in the code for most the things that I implemented. If you could make an extension of my fork, it would be great, but it needs a lot of refactoring of the jade extension, and I still don't understand the code submision policy. Who is the maintainer, you or @mquandalle? Who decides on what is accepted?

**Comment by mquandalle** [comment]
I’m the maintainer (and I try to listen to what everyone has to say).

**Comment by xiphias** [comment]
Sounds great, thanks! What I'd like to do is to add inline coffeescript support (which really can be an extension to the current code, as there's not much difference), and port the Meteor tutorial to this new coffee-jade extension. As I said, the main compatibility problem with anonymous helpers is that helpers can't call other helpers in Meteor (which can be hacked around, but I agree with @dalgard that it's not a good change).

**Comment by AlbinoGeek** [comment]
I would use this feature for no less than removing the "must create 100 tiny coffee / js files that simply have one 1-2 line method in them that does some thing much better solved inline."

Here are some such examples that could easily be replaced ...:

```coffee
  currentPath: (route) -> if route is Router.current().route.getName() then return 'active'
```

And the most obvious (href)

```coffee
unless currentUser
  a.ui.icon.warning.message(href=pathFor('login'))
    i.large.blue.key.icon
    .content
      .header Login Required
      p You're going to have to login first.
```

I also can't specify multiple things, such as classes (can't pass an array as the jade manual shows.)

Example:
```jade
  .projects
    .item(class={selected:selected,disabled:disabled}) ...
```

**Comment by laurentpayot** [comment]
Hi everyone. I come from a Python background and the Django template engine has an interesting approach to the issue of separation of concerns (MVC) versus templates expressiveness. They use a limited set of built-in filters like `length_is` or `join` : https://docs.djangoproject.com/en/1.8/ref/templates/builtins/#built-in-filter-reference
Would it be possible to implement a limited set of "filters" as well in Jade ? Just my two cents...

**Comment by xiphias** [comment]
Hey guys, as there was no merge, and I accept that this feature may not belong to vanilla jade package, I created a new package which you can use instantly: 

https://atmospherejs.com/xiphy/jade-coffee

Or just type in

meteor add xiphy:jade-coffee

It has JavaScript support as well if you're not a coffeescript fan. Please give it a try.
It's a replacement for mquandalle:jade, and I tried to stay compatible with it, so make sure that you don't add it in your project (or remove it).

**Comment by mquandalle** [comment]
@xiphias, your package doesn’t work on linux:

```
=> Errors prevented startup:                  
   
   While checking for xiphy:jade-coffee@0.1.0:
   error: No compatible binary build found for this package. Contact the package author and ask them to publish
   it for your platform.
```

**Comment by xiphias** [comment]
Thanks!
I'll delete xiphy:jade-coffee@0.1.0
Could you please try xiphy:jade-coffee@0.1.1 ?

**Comment by AlbinoGeek** [comment]
@xiphias Please use a Meteor Build Server to configure your package for architectures.

```
shadow@localhost$ meteor admin get-machine os.linux.x86_64 --minutes 5
user@buildserver# ... build
user@buildserver# meteor publish-for-arch ...
```

That's the recommended way of packaging, as it's a stable configuration that's maintained by them.  You can read more with the command `meteor help admin` or `meteor help admin get-machine`

You need only be logged in to your meteor account to use this command, not be a Meteor Administrator as the name would suggest (I think they mean "Maintainer", but I guess they failed a bit on the nomer.)

**Comment by xiphias** [comment]
Thanks very much, I'll do that and update here.
2015. okt. 23. 4:26 ezt írta ("Damon Blais" <notifications@github.com>):

**Comment by xiphias** [comment]
Thanks for the help, I found the problem with the build machines: I needed to add the right coffee-script npm dependency (it's interesting that after that I added it, it seems that the build machine is not necessary). Could you please try the newest version? (xiphy:jade-coffee@0.1.4)

**Comment by AlbinoGeek** [comment]
Hello @xiphias ,

The point of the build machines (that we get access to for free, which is a pretty good deal; as 5 minutes of AWS time every few days can add up to $10 or so, it's not a small cost for the end developers without servers to test on) is explained pretty well in the [Build Machines](https://github.com/meteor/meteor/wiki/Build-Machines) page.

And specifically this following section is important in this case: [Why is it important to use pre-configured build-machines.  Why shouldn't I just publish from my laptop?](https://github.com/meteor/meteor/wiki/Build-Machines#why-is-it-important-to-use-pre-configured-build-machines-why-shouldnt-i-just-publish-from-my-laptop)

**Comment by xiphias** [comment]
Hey, @AlbinoGeek,
I tried to do that, but I get this error message on the Linux build machine:
=> Errors while publishing build:
While publishing package build for xiphy:jade-coffee:
error: Cannot override existing build

Also I think that as I have a windows machine, and I haven't downloaded 4 cross-compilers on it, maybe meteor publish already does this. The only problem was that the error messages on the build machines weren't displayed when I was using meteor publish, so I didn't know what's the problem.
Also I couldn't find any detailed documentation on publishing packages using publish-with-arch.
I'm happy to give you maintainer rights to the packages and to my git repository if you give me your Meteor login though. I already added @mquandalle to the package maintainers just in case.

**Comment by AlbinoGeek** [comment]
Hi @xiphias ,

It turns out this is intentional (or at least known behaviour) as of posts such as [This particular closed issue on Meteor's GitHub](https://github.com/meteor/meteor/issues/4477#issuecomment-108110258) that explains that if a platform agnostic package already exists (i.e: Meteor detected no platform-specific code), then you may have an "all platforms" package published already (IsoPack)

Nothing wrong with that, and your builds work now on my projects :).  Good one.

**Comment by xiphias** [comment]
Thanks!
I just released a bug fix for the bug that I found today, feel free to
upgrade.
I think that the system is quite fun to work in, what is mostly missing for
me in my development is nice with/let support and multi-line quotes (as I
prefer having 0 helpers and as much code in my template as possible).

If you have experience in Meteor development and like the extension (but
don't want to get into the codebase), it would be nice if you could help
with writing about your experience and suggestions for people who are
starting to develop with jade-coffee, we could add it to the main page as
well. Anyways, have fun like me :)

**Comment by AlbinoGeek** [comment]
@xiphias Sure, any means we can communicate other than this thread?  I don't want other to be notified on our conversation about jade-coffee :).  mail via: (my github username) (at) gmail (dot) com , would work.

**Comment by AlbinoGeek** [comment]
@mquandalle We're merging in all changes from mquandalle:master branch back into xiphias:master , this PR might have many less differences soon.

**Comment by AlbinoGeek** [comment]
Right, I'll work towards fixing some of those 162 linter errors, heh.

**Comment by mquandalle** [comment]
If you have a modern version of `eslint`, `eslint --fix` will help you ;-)

**Comment by AlbinoGeek** [comment]
This still can't be merged, handler.js changes in 1.2 have not been implemented yet.  See: [Commit 215a966](https://github.com/xiphias/meteor-jade-coffee/commit/215a9667240ce29d2c708261ff4b3c3b495a69b6)  And thanks for the heads up on eslint.

**Comment by mquandalle** [comment]
Closing, because this is too old and can’t be merged.
